File: implementations/rust/src/ai.rs
use crate::types::*;
use crate::board::Board;
use crate::move_generator::MoveGenerator;
use crate::transposition_table::{TranspositionTable, BoundType, encode_move};
use std::time::Instant;

pub struct AI {
    move_generator: MoveGenerator,
    nodes_evaluated: u64,
    tt: TranspositionTable,
}

#[derive(Debug)]
pub struct SearchResult {
    pub best_move: Option<Move>,
    pub evaluation: i32,
    pub nodes: u64,
    pub time_ms: u128,
}

impl AI {
    pub fn new() -> Self {
        Self {
            move_generator: MoveGenerator::new(),
            nodes_evaluated: 0,
            tt: TranspositionTable::new(16),
        }
    }

    pub fn get_tt(&self) -> &TranspositionTable {
        &self.tt
    }

    pub fn get_tt_mut(&mut self) -> &mut TranspositionTable {
        &mut self.tt
    }

    pub fn find_best_move(&mut self, board: &Board, depth: u8) -> SearchResult {
        let start_time = Instant::now();
        self.nodes_evaluated = 0;
        
        let color = board.get_turn();
        let moves = self.move_generator.get_legal_moves(board, color);
        
        if moves.is_empty() {
            return SearchResult {
                best_move: None,
                evaluation: 0,
                nodes: 0,
                time_ms: 0,
            };
        }

        let mut best_move = moves[0].clone();
        let mut best_eval = if color == Color::White { i32::MIN } else { i32::MAX };

        for chess_move in &moves {
            let mut board_copy = board.get_state().clone();
            let mut test_board = Board::new();
            test_board.set_state(board_copy);
            test_board.make_move(chess_move);
            
            let evaluation = self.minimax(&test_board, depth - 1, i32::MIN, i32::MAX, color == Color::Black);
            
            if (color == Color::White && evaluation > best_eval) || 
               (color == Color::Black && evaluation < best_eval) {
                best_eval = evaluation;
                best_move = chess_move.clone();
            }
        }

        let elapsed = start_time.elapsed();
        SearchResult {
            best_move: Some(best_move),
            evaluation: best_eval,
            nodes: self.nodes_evaluated,
            time_ms: elapsed.as_millis(),
        }
    }

    fn minimax(&mut self, board: &Board, depth: u8, mut alpha: i32, mut beta: i32, maximizing: bool) -> i32 {
        self.nodes_evaluated += 1;

        // Probe transposition table
        let hash = board.get_state().hash;
        let original_alpha = alpha;

        if let Some(tt_entry) = self.tt.probe(hash) {
            if tt_entry.depth >= depth {
                match tt_entry.bound {
                    BoundType::Exact => {
                        return tt_entry.score;
                    }
                    BoundType::LowerBound => {
                        alpha = alpha.max(tt_entry.score);
                    }
                    BoundType::UpperBound => {
                        beta = beta.min(tt_entry.score);
                    }
                }
                if alpha >= beta {
                    return tt_entry.score;
                }
            }
        }

        if depth == 0 {
            let score = self.evaluate(board);
            self.tt.store(hash, 0, score, BoundType::Exact, None);
            return score;
        }

        let color = board.get_turn();
        let moves = self.move_generator.get_legal_moves(board, color);

        if moves.is_empty() {
            let score = if self.move_generator.is_in_check(board, color) {
                // Checkmate
                if maximizing { -100000 } else { 100000 }
            } else {
                // Stalemate
                0
            };
            self.tt.store(hash, depth, score, BoundType::Exact, None);
            return score;
        }

        if maximizing {
            let mut max_eval = i32::MIN;
            let mut best_move: Option<u16> = None;
            
            for chess_move in &moves {
                let mut board_copy = board.get_state().clone();
                let mut test_board = Board::new();
                test_board.set_state(board_copy);
                test_board.make_move(chess_move);
                
                let evaluation = self.minimax(&test_board, depth - 1, alpha, beta, false);
                
                if evaluation > max_eval {
                    max_eval = evaluation;
                    best_move = Some(encode_move(chess_move.from, chess_move.to));
                }
                alpha = alpha.max(evaluation);
                
                if beta <= alpha {
                    break; // Beta cutoff
                }
            }
            
            // Determine bound type
            let bound = if max_eval <= original_alpha {
                BoundType::UpperBound
            } else if max_eval >= beta {
                BoundType::LowerBound
            } else {
                BoundType::Exact
            };
            
            self.tt.store(hash, depth, max_eval, bound, best_move);
            max_eval
        } else {
            let mut min_eval = i32::MAX;
            let mut best_move: Option<u16> = None;
            
            for chess_move in &moves {
                let mut board_copy = board.get_state().clone();
                let mut test_board = Board::new();
                test_board.set_state(board_copy);
                test_board.make_move(chess_move);
                
                let evaluation = self.minimax(&test_board, depth - 1, alpha, beta, true);
                
                if evaluation < min_eval {
                    min_eval = evaluation;
                    best_move = Some(encode_move(chess_move.from, chess_move.to));
                }
                beta = beta.min(evaluation);
                
                if beta <= alpha {
                    break; // Alpha cutoff
                }
            }
            
            // Determine bound type
            let bound = if min_eval <= alpha {
                BoundType::LowerBound
            } else if min_eval >= beta {
                BoundType::UpperBound
            } else {
                BoundType::Exact
            };
            
            self.tt.store(hash, depth, min_eval, bound, best_move);
            min_eval
        }
    }

    fn evaluate(&self, board: &Board) -> i32 {
        let mut score = 0;

        for square in 0..64 {
            if let Some(piece) = board.get_piece(square) {
                let value = piece.piece_type.value();
                let position_bonus = self.get_position_bonus(square, piece.piece_type, piece.color, board);
                let total_value = value + position_bonus;
                
                score += if piece.color == Color::White { total_value } else { -total_value };
            }
        }

        score
    }

    fn get_position_bonus(&self, square: Square, piece_type: PieceType, color: Color, board: &Board) -> i32 {
        let file = square % 8;
        let rank = square / 8;
        let mut bonus = 0;

        // Center control bonus
        let center_squares = [27, 28, 35, 36]; // d4, e4, d5, e5
        if center_squares.contains(&square) {
            bonus += 10;
        }

        match piece_type {
            PieceType::Pawn => {
                // Pawn advancement bonus
                let advancement = if color == Color::White { rank } else { 7 - rank };
                bonus += (advancement * 5) as i32;
            },
            PieceType::King => {
                // King safety in opening/middlegame
                if !self.is_endgame(board) {
                    let safe_rank = if color == Color::White { 0 } else { 7 };
                    if rank == safe_rank && (file <= 2 || file >= 5) {
                        bonus += 20;
                    } else {
                        bonus -= 20;
                    }
                }
            },
            _ => {}
        }

        bonus
    }

    fn is_endgame(&self, board: &Board) -> bool {
        let mut piece_count = 0;
        let mut queen_count = 0;
        
        for square in 0..64 {
            if let Some(piece) = board.get_piece(square) {
                if piece.piece_type != PieceType::King && piece.piece_type != PieceType::Pawn {
                    piece_count += 1;
                    if piece.piece_type == PieceType::Queen {
                        queen_count += 1;
                    }
                }
            }
        }
        
        piece_count <= 4 || (piece_count <= 6 && queen_count == 0)
    }
}

File: implementations/zig/src/ai.zig
const std = @import("std");
const board = @import("board.zig");
const move_gen = @import("move_generator.zig");

pub const AI = struct {
    board_ref: *board.Board,
    last_evaluation: i32,
    nodes_searched: u64,

    pub fn init(board_ref: *board.Board) AI {
        return AI{
            .board_ref = board_ref,
            .last_evaluation = 0,
            .nodes_searched = 0,
        };
    }

    pub fn getBestMove(self: *AI, depth: u8) ?board.Move {
        self.nodes_searched = 0;

        const allocator = std.heap.page_allocator;
        var legal_moves = self.generateFilteredLegalMoves(allocator) catch return null;
        defer legal_moves.deinit(allocator);

        if (legal_moves.items.len == 0) return null;

        var best_move = legal_moves.items[0];
        var best_score: i32 = if (self.board_ref.white_to_move) -999999 else 999999;
        const maximizing = self.board_ref.white_to_move;

        for (legal_moves.items) |move| {
            const original_state = self.board_ref.*;
            self.board_ref.makeMove(move) catch continue;

            const score = self.minimax(depth - 1, -999999, 999999);

            self.board_ref.* = original_state;

            if (maximizing) {
                if (score > best_score) {
                    best_score = score;
                    best_move = move;
                }
            } else {
                if (score < best_score) {
                    best_score = score;
                    best_move = move;
                }
            }
        }

        self.last_evaluation = best_score;
        return best_move;
    }

    fn minimax(self: *AI, depth: u8, alpha: i32, beta: i32) i32 {
        self.nodes_searched += 1;

        const allocator = std.heap.page_allocator;
        var legal_moves = self.generateFilteredLegalMoves(allocator) catch return self.evaluatePosition();
        defer legal_moves.deinit(allocator);
        const current_color: board.PieceColor = if (self.board_ref.white_to_move) .White else .Black;
        const in_check = self.board_ref.isInCheck(current_color);

        if (legal_moves.items.len == 0) {
            if (in_check) {
                return if (self.board_ref.white_to_move) -100000 else 100000;
            }
            return 0;
        }

        if (depth == 0) {
            return self.evaluatePosition();
        }

        var current_alpha = alpha;
        var current_beta = beta;
        const maximizing = self.board_ref.white_to_move;

        if (maximizing) {
            var max_eval: i32 = -999999;

            for (legal_moves.items) |move| {
                const original_state = self.board_ref.*;
                self.board_ref.makeMove(move) catch continue;

                const eval = self.minimax(depth - 1, current_alpha, current_beta);

                self.board_ref.* = original_state;

                max_eval = @max(max_eval, eval);
                current_alpha = @max(current_alpha, eval);

                if (current_beta <= current_alpha) {
                    break; // Beta cutoff
                }
            }

            return max_eval;
        } else {
            var min_eval: i32 = 999999;

            for (legal_moves.items) |move| {
                const original_state = self.board_ref.*;
                self.board_ref.makeMove(move) catch continue;

                const eval = self.minimax(depth - 1, current_alpha, current_beta);

                self.board_ref.* = original_state;

                min_eval = @min(min_eval, eval);
                current_beta = @min(current_beta, eval);

                if (current_beta <= current_alpha) {
                    break; // Alpha cutoff
                }
            }

            return min_eval;
        }
    }

    pub fn evaluatePosition(self: *AI) i32 {
        var score: i32 = 0;

        // Material evaluation
        for (self.board_ref.squares, 0..) |piece, i| {
            if (piece) |p| {
                var piece_value = self.getPieceValue(p.piece_type);

                // Add positional bonuses
                piece_value += self.getPositionalBonus(p, @intCast(i));

                if (p.color == .White) {
                    score += piece_value;
                } else {
                    score -= piece_value;
                }
            }
        }

        return score;
    }

    fn getPieceValue(self: *AI, piece_type: board.PieceType) i32 {
        _ = self;
        return switch (piece_type) {
            .Pawn => 100,
            .Knight => 320,
            .Bishop => 330,
            .Rook => 500,
            .Queen => 900,
            .King => 20000,
        };
    }

    fn getPositionalBonus(self: *AI, piece: board.Piece, square: u8) i32 {
        _ = self;
        const file = square % 8;
        const rank = square / 8;
        var bonus: i32 = 0;

        // Center control bonus
        if ((file == 3 or file == 4) and (rank == 3 or rank == 4)) {
            bonus += 10;
        }

        // Pawn advancement bonus
        if (piece.piece_type == .Pawn) {
            if (piece.color == .White) {
                bonus += @as(i32, @intCast(rank)) * 5;
            } else {
                bonus += @as(i32, @intCast(7 - rank)) * 5;
            }
        }

        // King safety penalty for exposed king
        if (piece.piece_type == .King) {
            // Simplified king safety - penalize king in center during opening/middlegame
            if ((file >= 2 and file <= 5) and (rank >= 2 and rank <= 5)) {
                bonus -= 20;
            }
        }

        return bonus;
    }

    pub fn getLastEvaluation(self: *AI) i32 {
        return self.last_evaluation;
    }

    pub fn getNodesSearched(self: *AI) u64 {
        return self.nodes_searched;
    }

    fn generateFilteredLegalMoves(self: *AI, allocator: std.mem.Allocator) !std.ArrayList(board.Move) {
        var move_generator = move_gen.MoveGenerator.init(self.board_ref);
        var pseudo_moves = try move_generator.generateLegalMoves(allocator);
        defer pseudo_moves.deinit(allocator);

        var legal_moves = std.ArrayList(board.Move).empty;
        errdefer legal_moves.deinit(allocator);

        const current_color: board.PieceColor = if (self.board_ref.white_to_move) .White else .Black;
        for (pseudo_moves.items) |move| {
            const snapshot = self.board_ref.*;
            self.board_ref.makeMove(move) catch {
                self.board_ref.* = snapshot;
                continue;
            };

            if (!self.board_ref.isInCheck(current_color)) {
                try legal_moves.append(allocator, move);
            }

            self.board_ref.* = snapshot;
        }

        return legal_moves;
    }
};
